Services/Devices.cs

```diff
@@ -198,6 +209,15 @@ public void SetCurrentIotHub()
             this.log.Info("Selected active IoT Hub for devices", () => new { this.ioTHubHostName });
         }
 
+        private void SetupHub()
+        {
+            if (this.setupDone) return;
+
+            this.SetCurrentIotHub();
```

Comment:
Is this OK to be reentrant; i.e. is anything shared across threads?  It has a race condition for if methods are called in parallel.